### ISSUE TITLE
Contain text cursor inside block checkbox does not restrict the cursor when editing text

### ISSUE THREAD

**Issue by tbdalgaard** [WordPress/gutenberg#42659]
### Description

When the checkbox "Contain text cursor inside block" is checked normal keyboard navigation inside a block is restricted to the selected block.

But when a user edit the block content some unexpected results can be found. For example it is very easy to delete a paragraph block while editing the text of set paragraph.

### Step-by-step reproduction instructions

1. Make a new post.
2. If not set go to options, options, and click on the blocks tab.
3. Check the checkbox: "Contain text cursor inside block"
4. Close the options dialog.
5. Make a new paragraph and type.

	This is the first paragraph.
6. Press return. A new paragraph block is inserted.
7. type the following into this block:

	This is the second paragraph.
8. Move the cursor so it is at the beginning of the block. It should be placed to the left for the word This.
8. Press delete/backspace.
9. The second paragraph block is deleted and all content is moved into the first paragraph.

Here it would be easier for screenreader users if the block deletion only was possible when the navigation mode is active.

So if the user pressed delete as mentioned in the example above the result would be that nothing changed, because the editing mode is still active.

### Screenshots, screen recording, code snippet

No code can be pasted here.

### Environment info

Wordpress 6.0.1 Gutenberg: The integrated version inside Wordpress 6.01. Operativsystem and browser:
Mac OS 10.15.7 and Safari 15.6

### Please confirm that you have searched existing issues in the repo.

Yes

### Please confirm that you have tested with all plugins deactivated except Gutenberg.

No

**Comment by talldan** [comment]
That makes sense to me. Both enter and delete or backspace seem like they shouldn't create new blocks or delete blocks.

@alexstine what do you think about this?

**Comment by alexstine** [comment]
@talldan I think we have to keep the Enter key to create new blocks. Inserting a new block after seems like it would be tricky without a quick shortcut. However, I agree deleting the block would be super jarring. Do you have the time right now to open a PR? I believe most of that is done in writing flow use-arrow-nav hook. Happy to test, just do not have time to work on code ATM.

Thanks.

**Comment by tbdalgaard** [comment]
Would a compromise be the following:

If the checkbox is checked the following keyboard shortcuts would work when in editing mode and the cursor is inside a block:

- shift+enter: Make an empty paragraph block. Where as enter now makes it by default.
- shift-delete: Delete the actual paragraph and move contents upwards.

**Comment by alexstine** [comment]
@tbdalgaard Just a couple questions.

> shift-delete: Delete the actual paragraph and move contents upwards.

1. Would this be okay on Mac with no Delete key?

> shift+enter: Make an empty paragraph block. Where as enter now makes it by default.

2. I would rather not use this shortcut because generally Shift key signifies moving backwards. Enter truly does make sense in this context since you are moving forward, not backwards.

Thoughts?

**Comment by tbdalgaard** [comment]
> shift-delete: Delete the actual paragraph and move contents upwards.
Would this be okay on Mac with no Delete key?

Forgot the mention that I thought of backspace/delete here so both should work.

> shift+enter: Make an empty paragraph block. Where as enter now makes it by default.
I would rather not use this shortcut because generally Shift key signifies moving backwards. Enter truly does make sense in this context since you are moving forward, not backwards.
Thoughts?

I certainly understand where you are coming from, and I would agree if it wasn't that some blocks such as the Markdown block in the Jetpack plugin does make this possible, and it was much easier for me to work like this. But I understand why this shortcut was done in the first place. Perhaps this could stay as it is for now, but it would be very nice to find some documentation for screenreader users about this. That would have saved me many frustrations when I tried Gutenberg for a long time.

However, I will try to rethink the second suggestion about the enter key and do some tests and share my thoughts if that's ok.

**Comment by alexstine** [comment]
@tbdalgaard Check this: https://www.nvaccess.org/files/nvda/documentation/userGuide.html#SystemFocus

> The most common way of navigating around Windows with NVDA is to simply move the system focus using standard Windows keyboard commands, such as pressing tab and shift+tab to move forward and back between controls, pressing alt to get to the menu bar and then using the arrows to navigate menus, and using alt+tab to move between running applications.

As you can see, using Shift+Enter in this context would not fix our communication problem. In the context of the markdown block, I now understand why this is annoying for you. I think we still need to think about the effects on accessibility though. I use a screen reader, I would hate to have this changed at this point just due to what users have come to expect.

Thanks.

**Comment by tbdalgaard** [comment]
@alexstine 
You are right about the quote from the NVDA documentation. Note however that most of these rules are not defined when working in web environments, because there are not many screenreader manifactures that have tried to figure out some good standard ways. This is even harder because the developer of the web application can design it differently, and how they want.

A good example is when leaving a comment on a Facebook post: I recently noticed that pressing return/enter in the comment field will post the comment. There are no button that shows "Post comment" like how GitHub does it, and how should I as a user of a screenreader know that pressing enter/return will post the comment? What if I wanted to insert a hard line break instead of posting a comment on Facebook where enter/return normally would insert a hard line break and not press some defined, hidden post comment action?

The same is true for this issue inside the Gutenberg editor. How can I as a screenreader user guess that if I press delete at the beginning of a block the block itself is deleted, and the contents is moved up into the previous block?

I tried to write a few paragraphs and blocks without the Markdown-block I normally use, and found that it works quite good, when I understood how to navigate the editor and how to edit the current block. Here I really wish I have had some documentation that could explain how I was expected to use the Gutenberg editor.
We may need to think of a user experience that will work for most users. I understand why the editor behaves as it does, but perhaps we might think of making some notes for screenreader users instead of changing the behaviour inside the Gutenberg editor. I have searched for years, but it has ben very hard to find specific instructions if you are using a screenreader. If proper documentation was done this issue wouldn't have ben created since I would have known how to use the editor.

Another example:
I looked through the keyboard shortcut section in the Gutenberg plugin, and couldn't find any instructions about how to select multiple blocks noncontiguously. Is that even possible?

How can I get involved in communicating about documentation instead of functionality? Is this discussed on GitHub as well, or where should I look from here?

**Comment by alexstine** [comment]
@tbdalgaard Would you be open to meeting with me via Zoom? I am not sure how much I can help with this, but it sounds like you may have feedback very similar to my own. Gutenberg is a very custom app with a lot of custom odd things that screen reader users would not normally expect. The whole concept of a block editor I think is bad for accessibility. However, getting some good action points on what to focus on would be helpful. I can tell you for sure that the whole editor like a markdown block will not happen. It is also getting harder and harder to get other contributors to help out with some of these advanced accessibility fixes or even getting testing. There are only a few of us working on stuff like this.

Sounds like your problem you reported comes from custom interactions and things happening that you would not expect.

If you are open to meeting, you can find me on the Make WordPress Slack at alexstine.

CC'ing @annezazu as this is likely going to be a pain point as things keep growing in the editor. The more features, the more custom interactions for keyboard users.

Thanks.

**Comment by annezazu** [comment]
Thanks so much for flagging this, Alex, and Thomas for opening this issue. Getting as specific as possible with feedback makes advocating easier so do let me know what I can do if/when you all meet.

**Comment by tbdalgaard** [comment]
This would be very cool, but it should not change behaviour unless the checkbox is checked under preferences. I can see cases where other users may want this behaviour. I will be happy to test any changes if you want some feedback from me. I have used both VoiceOver on Mac for years and JAWS and NVDA on Windows.

**Comment by joedolson** [comment]
@alexstine Do you think there are any alternate solution to this available? I agree that this behavior with collapsing blocks can be undesirable (it's even more jarring if the second block is a heading, and it gets converted into part of the previous paragraph.)

I think it would be good if it was a bit more difficult to collapse blocks by accident.
